File: daemon/src/ingest/session_end.rs
//! `POST /hooks/session-end` handler.
//!
//! Marks `ended_at` (via `IngestEventRequest.mark_ended`) but never changes
//! `status` by itself — a `done`/`waiting-*` unresolved session stays
//! visible in the active queue after the process exits (D-06/D-07).

use axum::{extract::State, http::StatusCode, Json};
use serde_json::Value;
use std::sync::Arc;

use crate::{session_state::HookEvent, AppState, IngestEventRequest};

pub async fn session_end(
    State(state): State<Arc<AppState>>,
    Json(payload): Json<Value>,
) -> Result<StatusCode, StatusCode> {
    let session_id = payload
        .get("session_id")
        .and_then(Value::as_str)
        .ok_or(StatusCode::BAD_REQUEST)?
        .to_string();
    let cwd = payload
        .get("cwd")
        .and_then(Value::as_str)
        .map(str::to_string);
    let reason = payload
        .get("reason")
        .and_then(Value::as_str)
        .unwrap_or("unknown")
        .to_string();

    let request = IngestEventRequest {
        session_id,
        event: HookEvent::SessionEnd,
        cwd,
        notification_type: None,
        tool_name: None,
        timeline_summary: format!("session ended ({reason})"),
        payload_json: Some(payload.to_string()),
        is_error: false,
        first_prompt_text: None,
        mark_ended: true,
    };

    crate::ingest::dispatch_ingest_event(&state, request).await
}

File: daemon-ts/src/ingest/sessionEnd.ts
/**
 * `POST /hooks/session-end` handler.
 *
 * Marks `ended_at` (via `IngestEventRequest.markEnded`) but never changes
 * `status` by itself — a `done`/`waiting-*` unresolved session stays
 * visible in the active queue after the process exits. Mirrors
 * `daemon/src/ingest/session_end.rs`.
 */

import type { FastifyReply, FastifyRequest } from "fastify";
import type { Database as DatabaseType } from "better-sqlite3";

import { dispatchIngestEvent } from "./dispatch.js";
import type { IngestEventRequest } from "./dispatch.js";

export function makeSessionEndHandler(db: DatabaseType) {
  return async (req: FastifyRequest, reply: FastifyReply): Promise<void> => {
    const body = (req.body ?? {}) as Record<string, unknown>;
    const sessionId = typeof body.session_id === "string" ? body.session_id : undefined;
    if (!sessionId) {
      reply.code(400).send();
      return;
    }
    const cwd = typeof body.cwd === "string" ? body.cwd : null;
    const reason = typeof body.reason === "string" ? body.reason : "unknown";

    const request: IngestEventRequest = {
      sessionId,
      event: "SessionEnd",
      cwd,
      notificationType: null,
      toolName: null,
      timelineSummary: `session ended (${reason})`,
      payloadJson: JSON.stringify(body),
      isError: false,
      firstPromptText: null,
      markEnded: true,
    };

    dispatchIngestEvent(db, request);
    reply.code(200).send();
  };
}
